### ISSUE TITLE
Training PixelSeal from scratch?

### ISSUE THREAD

**Issue by JosTheBossX** [facebookresearch/videoseal#76]
Hello,

I did notice that this videoseal codebase has most of the things from the PixelSeal's paper if we want to train it from scratch. (other than the High-resolution adaptation) 

I did notice that the paper follows a 3 step training, 
1) The model trains on only BCE loss (with JND)
2) Then in second stage the discriminator loss (DLoss) is activated (with alpha decay)
3) Finetunining the model with th alpha as 0.2 + different types of attacks.

I have some questions:
a) In the stage 2, when we start the discriminator, does the alpha decay from 1->0.2 begins as soon as the DLoss is activated? or is there some delay between the alpha decay and the Dloss start?
b) In the stage 3, I noticed that in the paper it says it is finetuned on different attacks, does that mean that stage 1 and 2 were trained with clean watermarked images? i..e no noise layer?

I was trying to replicate by training PixelSeal from scratch, but for me the training PSNRs were only around 42. So I was also thinking are there any other things that might be missing in the videoseal codebase to train PixelSeal as per the paper?

Sorry I know these are a lot of questions, but I really liked the idea of adversarial-only training, without any other image losses, and wanted to try it from scratch.

Thanks :)

**Comment by pierrefdz** [comment]
Hello, thanks for waiting!

The PixelSeal paper (Section 4.2) describes a **three-stage training schedule**. Here is how they map to the released codebase.

### Stage 1: Message-only training (epochs 0–49)

- **Loss**: Only `L_msg` (BCE on message bits, `lambda_dec=1.0`)
- **Alpha (scaling_w)**: Fixed at `alpha_0 = 1.0` (watermark is very visible)
- **Discriminator**: OFF — the `--disc_start 50` flag zeros out the disc weight for all epochs < 50 via `adopt_weight()` in `videosealloss.py:25-31`
- **Augmentations**: All augmentations are active from epoch 0 (2 random augs per step from `all_augs.yaml`)
- **Perceptual loss**: `--lambda_i 0.0` "no perceptual loss".

**Purpose**: The model learns robust message embedding with a strong (visible) watermark. Bit accuracy should saturate quickly.

### Stage 2: Add discriminator + alpha decay (epochs 50–400)

- **Loss**: `L_msg + L_adv` (discriminator activates at epoch 50)
- **Alpha decay**: Cosine schedule from `1.0 → 0.2`, configured as:
  ```
  --scaling_w_schedule Cosine,scaling_min=0.2,start_epoch=200,epochs=200
  ```
  This means alpha stays at 1.0 for epochs 0–199, then cosine-decays from 1.0→0.2 over epochs 200–400.

**So to answer question (a)**: The discriminator activates at epoch 50, but the alpha decay doesn't start until epoch 200. There is a **150-epoch gap** between discriminator activation and the start of alpha decay. During epochs 50–199, the discriminator is active while alpha is still 1.0: the model is learning to hide a strong watermark before the watermark strength starts decreasing.

### Stage 3: Fine-tuning at final alpha (epochs 400–600)

- **Alpha**: Fixed at `0.2` (the cosine schedule has completed by epoch 400)
- **Loss**: Same as stage 2 (`L_msg + L_adv`)
- The remaining 200 epochs of training refine the model at the final watermark strength.

**To answer question (b)**: Stages 1 and 2 already use augmentations from the start. The paper's mention of ""potentially using different types of data, attacks, etc.". In the released model, the training uses all augmentations throughout.


For the change in PSNR, could you share exactly what you run? 
I remember that the high res adaptation led to higher PSNR, so it may be coming from this
